modules/Bio/EnsEMBL/DBSQL/BiotypeAdaptor.pm

```diff
@@ -0,0 +1,190 @@
+}
+
+=head2 _objs_from_sth
+
+  Arg [1]    : StatementHandle $sth
+  Example    : none
+  Description: PROTECTED implementation of abstract superclass method.
+               responsible for the creation of ProteinFeatures
+  Returntype : listref of Bio::EnsEMBL::Biotype objects
+  Exceptions : none
+
+=cut
+
+sub _objs_from_sth {
+  my ($self, $sth) = @_;
+
+  my ($dbID, $name, $object_type, $db_type, $attrib_type_id, $description, $biotype_group, $so_acc);
+
+  $sth->bind_columns(\$dbID, \$name, \$object_type, \$db_type, \$attrib_type_id, \$description, \$biotype_group, \$so_acc);
+
+  my @biotypes;
+
+  while($sth->fetch()) {
+    push( @biotypes,
+      my $feat = Bio::EnsEMBL::Biotype->new_fast( {
+         'dbID'           => $dbID,
+         'name'           => $name,
+         'object_type'    => $object_type,
+         'db_type'        => $db_type,
+         'attrib_type_id' => $attrib_type_id,
+         'description'    => $description,
+         'biotype_group'  => $biotype_group,
+         'so_acc'         => $so_acc,
+      } )
+    );
+  }
+
+  return \@biotypes;
+}
+
+
+=head2 fetch_by_name_object_type
+
+  Arg [1]    : String $name
+               The name of the biotype to retrieve
+  Arg [2]    : String $object_type
+               The object type of the biotype to retrieve (gene or transcript)
+  Example    : $biotype = $biotype_adaptor->fetch_by_name_object_type('gene', 'protein_coding');
+  Description: Retrieves a biotype object from the database via its combined key (name, object_type).
+               If the Biotype requested does not exist in the database, a new Biotype object is
+               created with the provided name and object_type to be returned.
+  Returntype : Bio::EnsEMBL::Biotype
+  Exceptions : none
+
+=cut
+
+sub fetch_by_name_object_type {
+  my ($self, $name, $object_type) = @_;
+
+  my $constraint = "b.name = ? AND b.object_type = ?";
+  $self->bind_param_generic_fetch($name, SQL_VARCHAR);
+  $self->bind_param_generic_fetch($object_type, SQL_VARCHAR);
+  my ($biotype) = @{$self->generic_fetch($constraint)};
+
+  if (!defined $biotype) {
+    $biotype = Bio::EnsEMBL::Biotype->new(
+          -NAME          => $name,
+          -OBJECT_TYPE   => $object_type,
+      )
+  }
+
+  return $biotype;
+}
+
+=head2 fetch_all_by_object_type
+
+  Arg [1]    : String $object_type
+               listref of $sources
+               The object_type of the biotypes to retrieve (gene or transcript).
+  Example    : $biotypes = $biotype_adaptor->fetch_all_by_object_type('gene');
+  Description: Retrieves an array reference of biotype objects from the database.
+  Returntype : listref of Bio::EnsEMBL::Biotype objects or empty list
+  Exceptions : none
+
+=cut
+
+sub fetch_all_by_object_type {
+  my ($self, $object_type) = @_;
+
+  my $constraint = "b.object_type = ?";
+  $self->bind_param_generic_fetch($object_type, SQL_VARCHAR);
+  my @biotypes = @{$self->generic_fetch($constraint)};
```

Comment:
What happens here if the query returns no rows? Might be better not to deref automatically and either generate an exception or return []

Reply:
Added a warning if no rows retrieved.
Empty array returned.

Reply:
I think you've got it covered. Would you mind putting in a test to prove that it fails gracefully? A proper exception is better than a warning which may go unnoticed. Does this code need to fail quietly?

Reply:
Added a test to check the warning and return value ([] in case of warning).
I suppose the warning can be an exception instead... Either a full replacement (if empty return value) or just like the exception in Bio::EnsEMBL::Biotype::object_type()
What do you think would be better?

Reply:
I prefer my code to fail loudly and early, as that greatly speeds debugging, but hard failures upset downstream users who have relied on the silent failures. It's your call.